File: TypeScript/inheritance.ts
// Base interface for Animal
interface Animal {
    name: string;
    sound: string;
    makeSound(): void;
}

// Base class implementing the Animal interface
class BaseAnimal implements Animal {
    constructor(public name: string, public sound: string) { }

    makeSound() {
        console.log(`${this.name} says ${this.sound}`);
    }
}

// Derived class Dog, extending from BaseAnimal
class Dog extends BaseAnimal {
    constructor(name: string) {
        super(name, "Woof!");
    }
}

// Derived class Cat, extending from BaseAnimal
class Cat extends BaseAnimal {
    constructor(name: string) {
        super(name, "Meow!");
    }
}

// Usage example
const dog = new Dog("Buddy");
const cat = new Cat("Cute");

dog.makeSound(); // Output: Buddy says Woof!
cat.makeSound(); // Output: Whiskers says Meow!

File: Python/Inheritance.py
class Animal:
    def __init__(self, name):
        self.name = name

    def make_sound(self):
        return "Some generic sound"

# single level inheritance
class Dog(Animal):
    def make_sound(self):
        return "Woof!"

class Cat(Animal):
    def make_sound(self):
        return "Meow!"

# Creating instances of derived classes
dog = Dog("Buddy")
cat = Cat("Whiskers")

print(dog.name)  # Output: Buddy
print(dog.make_sound())  # Output: Woof!

print(cat.name)  # Output: Whiskers
print(cat.make_sound())  # Output: Meow!
